### PR TITLE
proof: add decimal display as TLV field to meta reveal

### PR DESCRIPTION
**Pull request by guggero**
Minimal implementation that fixes https://github.com/lightninglabs/taproot-assets/issues/956.
Also fixes https://github.com/lightninglabs/taproot-assets/issues/997.

### CODE REVIEW

File: proof/meta.go [hunk 1]
```diff
@@ -66,26 +66,9 @@ var (
...
-	// invalid.
-	ErrDecDisplayInvalid = errors.New("invalid decimal display")
-
 	// ErrDecDisplayTooLarge is returned if the decimal display value
 	// exceeds the current limit.
 	ErrDecDisplayTooLarge = errors.New("decimal display too large")
-
-	// ErrDecDisplayInvalidType is returned if the value in a JSON object
-	// assigned to the decimal display key is not a uint32.
-	ErrDecDisplayInvalidType = errors.New("decimal display JSON field is " +
```
Comment (Roasbeef) [inline proof/meta.go:83]: Nice to get rid of all these error cases that no longer apply!

File: proof/meta_test.go [hunk 2]
```diff
@@ -59,6 +62,24 @@ func TestValidateMetaReveal(t *testing.T) {
 			},
 			expectedErr: nil,
 		},
+		{
+			name: "invalid decimal display",
+			reveal: &MetaReveal{
+				Type:           MetaJson,
+				Data:           []byte(`{"key": "value"}`),
+				DecimalDisplay: 999,
```
Comment (Roasbeef) [inline proof/meta_test.go:70]: 👍

File: proof/meta_test.go [hunk 3]
```diff
@@ -78,3 +99,137 @@ func TestValidateMetaReveal(t *testing.T) {
...
+			require.NoError(tt, err)
+
+			require.Equal(tt, tc.reveal, decoded)
+		})
+	}
+}
+
+// TestDecodeOldMetaReveal tests decoding an old MetaReveal encoding that does
+// not include the decimal display.
+func TestDecodeOldMetaReveal(t *testing.T) {
```
Comment (Roasbeef) [inline proof/meta_test.go:188]: Nice client compat tests 👍

**Review by Roasbeef — APPROVED** [review]
LGTM 💃🏾

File: rpcserver.go [hunk 4]
```diff
@@ -552,15 +552,13 @@ func (r *rpcServer) MintAsset(ctx context.Context,
 		// If a custom decimal display was requested, add that to the
 		// metadata and re-validate it.
 		if metaType == proof.MetaJson && req.Asset.DecimalDisplay != 0 {
```
Comment (jharveyb) [inline rpcserver.go:554]: I think this condition now needs to change and not require the `metaType` to be JSON?

File: itest/asset_meta_test.go [hunk 5]
```diff
@@ -149,10 +149,10 @@ func testMintAssetWithDecimalDisplayMetaField(t *harnessTest) {
 	}
 
 	// Manually update the requested metadata and compute the expected hash.
-	updatedMeta, err := mintedMeta.SetDecDisplay(firstAsset.DecimalDisplay)
+	err := mintedMeta.SetDecDisplay(firstAsset.DecimalDisplay)
```
Comment (jharveyb) [inline itest/asset_meta_test.go:152]: Should also cover non-JSON metadata to make sure the dec display gets passed through the minter (and then list assets) correctly.

**Review by jharveyb — COMMENTED** [review]
Looks good overall! Def cleaner than what we had before.

Added comments + I think we want to add this field to the `AssetMeta` proto as well + pipe it through there? If it's actually going to be a TLV field.

https://github.com/lightninglabs/taproot-assets/blob/5236deae2e3b3c73610063062643c5f32ab10ca4/taprpc/taprootassets.proto#L163

**Comment by jharveyb** [comment]
Also, I don't see the utility of adding this field as JSON in the metadata any more?

IIRC the actual user request was 'it shows up in RPC responses', and if its also discoverable from decoding proofs then it seems redundant. AFAICT we never read it after setting it.

**Comment by Roasbeef** [comment]
> Also, I don't see the utility of adding this field as JSON in the metadata any more?

Good point, we can just leave it out there.

**Comment by Roasbeef** [comment]
> I think we want to add this field to the AssetMeta proto as well 

Technically it's part of the the metal reveal, which contains the raw asset meta, and now these additional bytes. I guess it's useful though to have it there if people want to use `FetchAssetMeta` to query the decimal value.

**Comment by jharveyb** [comment]
> > I think we want to add this field to the AssetMeta proto as well
> 
> Technically it's part of the the metal reveal, which contains the raw asset meta, and now these additional bytes. I guess it's useful though to have it there if people want to use `FetchAssetMeta` to query the decimal value.

Just realized it's now also input for `MetaReveal.MetaHash()` - so definitely 'consensus' if we're going to leave that as is.

**Review by Roasbeef — CHANGES_REQUESTED** [review]
Discussed offline, and concluded that this approach was a bit too slick to be true:
  * Old clients won't know of this meta reveal field. 
  * They'll parse the meta reveal w/o the new field, then go to check against the meta hash.
  * That fails as they didn't include the portion they didn't know about.

File: proof/meta_test.go [hunk 6]
```diff
@@ -59,6 +62,24 @@ func TestValidateMetaReveal(t *testing.T) {
...
+			},
+			expectedErr: ErrDecDisplayTooLarge,
+		},
+		{
+			name: "correct decimal display",
+			reveal: &MetaReveal{
+				Type:           MetaJson,
+				Data:           []byte(`{"key": "value"}`),
+				DecimalDisplay: 8,
+			},
```
Comment (GeorgeTsagk) [inline proof/meta_test.go:80]: could also check against Opaque data type and also with 0 dec display

**Review by GeorgeTsagk — COMMENTED** [review]
So the purpose here was to drop this field to the `MetaReveal` level in order to support dec display for custom meta data?

**Comment by guggero** [comment]
> Discussed offline, and concluded that this approach was a bit too slick to be true:
> 
>     * Old clients won't know of this meta reveal field.
> 
>     * They'll parse the meta reveal w/o the new field, then go to check against the meta hash.
> 
>     * That fails as they didn't include the portion they didn't know about.

Ah, yes, nice catch.

Adding a summary of my comments made offline here for reference:

I see two possibilities for something we could do now:
1. don't add the decimal display to the meta hash (e.g. don't commit to it, just put it there as informational piece of data). not super nice and probably opens the door for shenanigans
2. add a "feature bit" to the meta type that indicates the meta data is nested TLV. then old nodes just read in the meta data as a blob and arrive at the correct hash. and new nodes would see the feature bit and do a second round of TLV decode on the data, being able to read and interpret extra fields that we put in there (the "actual" raw meta data being one of the sub fields). should work but feels veeery hacky...

I guess doing the two-level TLV encoding would qualify as "carveout/placeholder for future fields" as described in the issue #997. it just feels very hacky if we need to describe it in the BIP as "two levels of TLV encoding, depending on a feature bit"... but happy to go ahead and implement it, if we think it's the way to go

File: rpcserver.go (hunk 4 above)
Comment (guggero) [reply rpcserver.go]: Good point! Fixed this and a couple of other occurrences of similar checks.

**Comment by guggero** [comment]
Okay, the current approach doesn't work and any quick hacks are too hacky. I'll implement the proper carve out to be able to do this properly in the next major version in an upcoming PR.

**Pull request closed without merging**